plugin/gcp/instance.py

```diff
@@ -0,0 +1,239 @@
+from plugin.gcp.service import blocking
+
+
+class Instance(GoogleCloudPlatform):
+    ACCESS_CONFIG = 'External NAT'
+    ACCESS_CONFIG_TYPE = 'ONE_TO_ONE_NAT'
+    NETWORK_INTERFACE = 'nic0'
+    STANDARD_MACHINE_TYPE = 'n1-standard-1'
+
+    def __init__(self,
+                 config,
+                 logger,
+                 instance_name,
+                 image=None,
+                 machine_type=None,
+                 startup_script=None,
+                 external_ip=False,
+                 tags=[]):
+        """
+        Create Instance object
+
+        :param config: gcp auth file
+        :param logger: logger object
+        :param instance_name: name of the instance
+        :param image: image id in Google Cloud Platform
+        :param machine_type: machine type on GCP, default None
+        :param startup_script: shell script text to be run on instance startup,
+        default None
+        :param external_ip: boolean external ip indicator, default False
+        :param tags: tags for the instance, default []
+        """
+        super(Instance, self).__init__(config, logger)
+        self.project = config['project']
+        self.zone = config['zone']
+        self.name = utils.get_gcp_resource_name(instance_name)
+        self.image = image
+        self.machine_type = machine_type \
+            if machine_type else self.STANDARD_MACHINE_TYPE
+        self.network = config['network']
+        self.startup_script = startup_script
+        self.tags = tags
+        self.externalIP = external_ip
+
+    @blocking(True)
+    def create(self):
+        """
+        Create GCP VM instance with given parameters.
+        Zone operation.
+
+        :return: REST response with operation responsible for the instance
+        creation process and its status
+        :raise: GCPError if there is any problem with startup script file:
+        e.g. the file is not under the given path or it has wrong permissions
+        """
+        self.logger.info('Create instance')
+        body = self.to_dict()
+
+        if self.startup_script is not None:
+            try:
+                with open(self.startup_script, 'r') as script:
+                    item = {
+                        'key': 'startup-script',
+                        'value': script.read()
+                    }
+                    body['metadata']['items'].append(item)
+            except IOError as e:
+                self.logger.error(str(e))
+                raise GCPError(str(e))
+
+        return self.compute.instances().insert(
+            project=self.project,
+            zone=self.zone,
+            body=body).execute()
+
+    @blocking(True)
+    def delete(self):
+        """
+        Delete GCP instance.
+        Zone operation.
+
+        :return: REST response with operation responsible for the instance
+        deletion process and its status
+        """
+        self.logger.info('Delete instance')
+        return self.compute.instances().delete(
```

Comment:
The error thrown by GCP is really very detailed and I don't see need of checking whether the instance exists before. All tasks raise a non-recovery exception in case of GCPError.